### PR TITLE
[Feature]Add cugraph local storage support to DGL

### PR DESCRIPTION
**Pull request by VibhuJawa**
## Description

This PR closes https://github.com/dmlc/dgl/issues/4672 by untegrating cugraph local storage into DGL.  

Please note that `Nvidia Internal Review` for this PR done at: https://github.com/rapidsai/dgl/pull/16

Accuracy Results on https://github.com/dmlc/dgl/tree/master/examples/pytorch/rgcn-hetero by using the same defaults mentioned there so **hyper parameter tuning ins not done.**  but we are still in a percentage point of the avg runs . 

| dataset | Avg(cugraph, 18 runs) | Avg(DGL, 5 runs) |
|---------|-----------------------|------------------|
| aifb    | 93.81                 | 94.44            |
| bgs     | 90.04                 | 91.04            |
| mutag   | 66.17                 | 67.37            |

This PR enables cugraph storage by using duck-typing by following the `Extending DGL Graph Sampling Pipeline
` RFC (See [link](https://hackmd.io/6PJixJj7Ssy19I-N9CV1Ag?view))  so all one needs to do is to switch the graph object from `DGLGraph` object to `CuGraphStorage` object. 

See below:
```diff

+from dgl.contrib.cugraph.convert import cugraph_storage_from_heterograph
+cugraph_g = cugraph_storage_from_heterograph(dgl_g)

sampler = dgl.dataloading.NeighborSampler(
        [15, 10, 5], prefetch_node_feats=['feat'], prefetch_labels=['label'])

train_dataloader = dgl.dataloading.DataLoader(
- dgl_g, 
+ cugraph_g,
train_idx, 
sampler, 
device=device, 
batch_size=1024,
shuffle=True,
drop_last=False, 
num_workers=0)

```

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR

## Features Added
-  [x] cugraph Homogeneous Graph
-  [x]  cugraph Heterogeneous Graph
-  [x] Single GPU cugraph support
-  [x] Multi GPU cugraph support
-  [x] End to End Examples 
-  [x] from hetrogenous_graph

##  Tests: 
-  [x] Homogeneous Graph Sampling
-  [x] Heterogeneous Graph Sampling
-  [x] from hetrogenous_graph

## Out of scope:
<del> renumber_vertices_by_type added in API
<del>  to_hetrogenous_graph
<del> edge_subgraph
<del> subgraph

### CODE REVIEW

File: examples/cugraph-pytorch/cugraph-local/rgcn-hetero/model.py [hunk 1]
```diff
@@ -0,0 +1,569 @@
+"""RGCN layer implementation"""
+# File same as: https://github.com/dmlc/dgl/blob/master/examples/pytorch/rgcn-hetero/model.py
+from collections import defaultdict
```
Comment (VibhuJawa) [inline examples/cugraph-pytorch/cugraph-local/rgcn-hetero/model.py:3]: No change made in this file.

File: examples/cugraph-pytorch/cugraph-local/rgcn-hetero/entity_classify_mb.py [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+Paper: https://arxiv.org/abs/1703.06103
+Reference Code: https://github.com/tkipf/relational-gcn
+"""
+# Modified the file at:
+# https://github.com/dmlc/dgl/blob/master/examples/pytorch/rgcn-hetero/entity_classify_mb.py
+# To work with cugraph_storage
+# Only change is:
+# from dgl.contrib.cugraph.convert import cugraph_storage_from_heterograph
+# g = cugraph_storage_from_heterograph(g)
+
```
Comment (VibhuJawa) [inline examples/cugraph-pytorch/cugraph-local/rgcn-hetero/entity_classify_mb.py:11]: See above note

**Comment by VibhuJawa** [comment]
CC: @jermainewang , @TristonC

**Comment by VibhuJawa** [comment]
@jermainewang , It would be amazing if you help this PR reviewed . Thanks again for the help.
